src/inmanta/agent/agent.py

```diff
@@ -1491,84 +1598,91 @@ async def on_disconnect(self) -> None:
-            # TODO deal with this lock  -> not sure how
-            async with self._resource_loader_lock.get(resource_type):
-                result: protocol.Result = await self._client.get_source_code(environment, version, resource_type)
-                if result.code == 200 and result.result is not None:
-                    sync_client = SyncClient(client=self._client, ioloop=self._io_loop)
-                    LOGGER.debug("Installing handler %s version=%d", resource_type, version)
-                    requirements: set[str] = set()
-                    sources: list["ModuleSource"] = []
-                    for source in result.result["data"]:
-                        sources.append(
-                            ModuleSource(
-                                name=source["module_name"],
-                                is_byte_code=source["is_byte_code"],
-                                hash_value=source["hash"],
-                                _client=sync_client,
-                            )
+            result: protocol.Result = await self._client.get_source_code(environment, version, resource_type)
+            if result.code == 200 and result.result is not None:
+                sync_client = SyncClient(client=self._client, ioloop=self._io_loop)
+                LOGGER.debug("Installing handler %s version=%d", resource_type, version)
+                requirements: set[str] = set()
+                sources: list["ModuleSource"] = []
+                # Encapsulate source code details in ``ModuleSource`` objects
+                for source in result.result["data"]:
+                    sources.append(
+                        ModuleSource(
+                            name=source["module_name"],
+                            is_byte_code=source["is_byte_code"],
+                            hash_value=source["hash"],
+                            _client=sync_client,
                         )
-                        requirements.update(source["requirements"])
+                    )
+                    requirements.update(source["requirements"])
 
-                    if pip_config is None:
-                        # TODO: what if this fails?
-                        pip_config = await self._get_pip_config(environment, version)
+                if pip_config is None:
+                    # TODO: what if this fails?
+                    pip_config = await self._get_pip_config(environment, version)
 
-                    resource_install_specs.append(
-                        ResourceInstallSpec(resource_type, version, pip_config, list(requirements), sources)
-                    )
-                else:
-                    # TODO what if server call fails ?
-                    pass
+                resource_install_specs.append(
+                    ResourceInstallSpec(resource_type, version, pip_config, list(requirements), sources)
+                )
+            else:
+                # TODO what if server call fails ?
```

Comment:
TODO alert

Reply:
I replicated the initial behavior: mark resources for which we can't get the model source or the pip config as failed